simplemma/strategies/dictionaries/trie_directory_factory.py

```diff
@@ -0,0 +1,161 @@
+from platformdirs import user_cache_dir
+
+from simplemma.strategies.dictionaries.dictionary_factory import (
+    SUPPORTED_LANGUAGES,
+    DictionaryFactory,
+    _load_dictionary_from_disk,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class TrieWrapDict(MutableMapping):
+    """Wrapper around BytesTrie to make them behave like dicts."""
+
+    def __init__(self, trie: BytesTrie):
+        self._trie = trie
+
+    def __getitem__(self, item):
+        return self._trie[item.decode()][0]
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError
+
+    def __delitem__(self, key):
+        raise NotImplementedError
+
+    def __iter__(self):
+        for key in self._trie.iterkeys():
+            yield key.encode()
+
+    def __len__(self):
+        return len(self._trie)
+
+
+class TrieDictionaryFactory(DictionaryFactory):
+    """Memory optimized DictionaryFactory backed by MARISA-tries.
+
+    This dictionary factory creates dictionaries, which are backed by a
+    MARISA-trie instead of a dict, to make them consume very little
+    memory compared to the DefaultDictionaryFactory. Trade-offs are that
+    lookup performance isn't as good as with dicts.
+    """
+
+    __slots__: List[str] = []
+
+    def __init__(
+        self,
+        cache_max_size: int = 8,
+        use_disk_cache: bool = True,
+        disk_cache_dir: Optional[str] = None,
+    ) -> None:
+        """Initialize the TrieDictionaryFactory.
+
+        Args:
+            cache_max_size (int): The maximum number dictionaries to
+                keep in memory. Defaults to `8`.
+            use_disk_cache (bool): Whether to cache the tries on disk to
+                speed up loading time. Defaults to `True`.
+            disk_cache_dir (Optional[str]): Path where the generated
+                tries should be stored in. Defaults to a Simplemma-
+                specific subdirectory of the user's cache directory.
+        """
+
+        if disk_cache_dir:
+            self._cache_dir = Path(disk_cache_dir)
+        else:
+            self._cache_dir = Path(user_cache_dir("simplemma")) / "marisa_trie"
+        self._use_disk_cache = use_disk_cache
+        self._get_dictionary = lru_cache(maxsize=cache_max_size)(
+            self._get_dictionary_uncached
+        )
+
+    def _get_pickled_dict_path(self, lang: str) -> Path:
+        """Return the file path of the pickled dictionary."""
+        return Path(__file__).parent / Path(f"data/{lang}.plzma")
+
+    def _get_pickled_dict_hash(self, lang: str) -> str:
+        """Get the SHA256 hash of a pickled dict file."""
+        hasher = hashlib.sha256()
+        with self._get_pickled_dict_path(lang).open("rb") as fp:
+            while True:
+                chunk = fp.read(4096)
+                if not chunk:
+                    break
+                hasher.update(chunk)
+        return hasher.hexdigest()
+
+    def _cache_is_current(self, lang: str) -> bool:
```

Comment:
What's the scenario in which the in-disk trie won't match the in-memory one?

Reply:
This method is for checking whether the tries cached on disk still match the pickled bytestring dictionaries. They don't match anymore whenever a new version of Simplemma gets released which includes updated pickled bytestring dictionaries, resulting in the need to regenerate the cached tries.